jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java

```diff
@@ -448,63 +516,52 @@ private void terminateResponse(HttpExchange exchange, Result result)
     }
 
     /**
-     * Resets this {@link HttpReceiver} state.
+     * Resets the state of this HttpReceiver.
      * <p>
      * Subclasses should override (but remember to call {@code super}) to reset their own state.
      * <p>
      * Either this method or {@link #dispose()} is called.
      */
     protected void reset()
     {
-        contentListeners = null;
-        destroyDecoder(decoder);
-        decoder = null;
+        cleanup();
     }
 
     /**
-     * Disposes this {@link HttpReceiver} state.
+     * Disposes the state of this HttpReceiver.
      * <p>
      * Subclasses should override (but remember to call {@code super}) to dispose their own state.
      * <p>
      * Either this method or {@link #reset()} is called.
      */
     protected void dispose()
     {
-        destroyDecoder(decoder);
-        decoder = null;
+        cleanup();
     }
 
-    private static void destroyDecoder(ContentDecoder decoder)
+    private void cleanup()
     {
-        if (decoder instanceof Destroyable)
-        {
-            ((Destroyable)decoder).destroy();
-        }
+        contentListeners = null;
+        if (decoder != null)
+            decoder.destroy();
+        decoder = null;
+        demand = 0;
+        stalled = false;
     }
 
     public boolean abort(HttpExchange exchange, Throwable failure)
     {
         // Update the state to avoid more response processing.
         boolean terminate;
-        out:
         while (true)
         {
             ResponseState current = responseState.get();
-            switch (current)
+            if (current == ResponseState.FAILURE)
+                return false;
+            if (updateResponseState(current, ResponseState.FAILURE))
             {
-                case FAILURE:
-                {
-                    return false;
-                }
-                default:
-                {
-                    if (updateResponseState(current, ResponseState.FAILURE))
-                    {
-                        terminate = current != ResponseState.TRANSIENT;
-                        break out;
-                    }
-                    break;
-                }
+                terminate = current != ResponseState.TRANSIENT;
```

Comment:
comment to explain why transients don't need to be terminated here would be good?   I guess this is partially answering my question above about who else can leave TRANSIENT